### ISSUE TITLE
JUnit5: @Testable is ignored if org.junit.jupiter:junit-jupiter-engine is missing

### ISSUE THREAD

**Issue by OLibutzki** [eclipse/xtext-xtend#519]
Hi everyone,

I have a Xtend class which is annotated by @org.junit.platform.commons.annotation.Testable. Unfortunately I cannot choose "Run As" | "JUnit Test".

After adding the org.junit.jupiter:junit-jupiter-engine dependency "Run As" | "JUnit Test" is displayed.

As I have a custom test engine I do not need and want any Jupiter dependencies. 

org.junit.platform:junit-platform-engine should be sufficient.

Kind regards
Oliver

**Comment by cdietrich** [comment]
afaik it is intentional that you need to add a engine.
see the discussion at https://github.com/eclipse/xtext-eclipse/issues/653

**Comment by cdietrich** [comment]
besides that org.eclipse.xtext.xbase.ui.launching.JavaElementDelegateJunitLaunch.containsElementsSearchedFor(IFile) searches for methods named test* or having `@Test` to determine the launchability

**Comment by cdietrich** [comment]
=> could you provide a minimal example? 

=> in my code if i dont have the engine / (a engine?) then the code above wont be called at all.

**Comment by cdietrich** [comment]
example pom with no engine at all. the question: what other critera decide (on jdt side)

```
<project xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<groupId>a.demo</groupId>
	<artifactId>a.demo</artifactId>
	<version>0.0.1-SNAPSHOT</version>

	<dependencies>
		<dependency>
			<groupId>org.junit.jupiter</groupId>
			<artifactId>junit-jupiter-engine</artifactId>
			<version>5.1.0</version>
			<exclusions>
				<exclusion>
					<groupId>org.junit.platform</groupId>
					<artifactId>junit-platform-engine</artifactId>
				</exclusion>
			</exclusions>
		</dependency>
		<dependency>
			<groupId>org.junit.jupiter</groupId>
			<artifactId>junit-jupiter-api</artifactId>
			<version>5.1.0</version>
		</dependency>
		<dependency> 
			<groupId>org.eclipse.xtext</groupId>
			<artifactId>org.eclipse.xtext.xbase.lib</artifactId>
			<version>2.14.0</version>
		</dependency>
	</dependencies>
</project>
```

**Comment by cdietrich** [comment]
see `org.eclipse.jdt.internal.junit.launcher.JUnit3TestFinder.isTest(IType)`

**Comment by OLibutzki** [comment]
Here is the example: https://github.com/OLibutzki/Xtend519

There is a custom engine *MyTestEngine* which is registered via *META-INF/services/org.junit.platform.engine.TestEngine*. The engine just adds the selected class as test in the discovery phase and marks this class as executed successfully in the execution phase.

Moreover there is an annotation wich is meta-annotated by *Testable* in order to tell the IDE to identify the test class candidates.

I added two test classes: *MyJavaTest.java* and *MyXtendTest.xtend*. *MyJavaClass* can be launched as JUnit Test, but for the Xtend class this option is not displayed. Anyway, I can run it a JUnit Test if I select the generated Java class *MyXtendTest.java*.

**Comment by cdietrich** [comment]
yes as i said we consider only stuff with testxxx and `@Test` as testable
=> we would first need to specify other cases

**Comment by OLibutzki** [comment]
Ok, but JUnit5 does not declare any special treatment of @Test or testxxx. For IDE support there is just the Testable annotation which marks a type or method as test candidate.

In my project (I cannot reproduce it in the small example) I can execute an Xtend class which is called *MyFirstTest* and which has one method called *step1* as JUnit test. I just don't know where the difference is...

**Comment by OLibutzki** [comment]
I just investigated this further and `JavaElementDelegateJunitLaunch` behaves strange as `element.exists()` always returns false, if the generated java file is not opened in an editor. So all the checks concerning `@Test` or `testxxx` are not reachable (unless the generated java file is opened).

I use Eclipse Photon.

**Comment by cdietrich** [comment]
i did not see that is my debugging. are you sure the xtend-gen folder is a proper source folder?

**Comment by OLibutzki** [comment]
> i did not see that is my debugging. are you sure the xtend-gen folder is a proper source folder?
Sorry, my bad.

Anyway, *JavaElementDelegateJunitLaunch* is consulted when copening the context menu in the Package Explorer. When performing a right-click in the editor the *JavaElementDelegateJunitLaunch* seems not to be consulted.

Is there a "fallback" to the jdt implementation when opening the context menu in the editor?

It turns out that this seems to be a combination of issues:

1. *JavaElementDelegateJunitLaunch* is not consulted for building the editor's context menu
1. *JavaElementDelegateJunitLaunch* does not respect `@Testable`
1. The context menu for the editor does not respect `@Testable` unless I add the org.junit.jupiter:junit-jupiter-engine

I'm getting crazy more and more ;-)

**Comment by cdietrich** [comment]
me too. the junit 5 support was not really addressed in all the context menu stuff.

**Comment by cdietrich** [comment]
see https://github.com/eclipse/xtext-eclipse/issues/768 too

**Comment by cdietrich** [comment]
did some more debugging. for the open editor `org.eclipse.xtext.xbase.ui.launching.JavaElementDelegateAdapterFactory.getAdapter(Object, Class)` is used.

**Comment by cdietrich** [comment]
=> in your case

```
 <or>
                              <test property="org.eclipse.jdt.core.hasTypeOnClasspath" value="junit.framework.Test"/>
                              <test property="org.eclipse.jdt.core.hasTypeOnClasspath" value="org.junit.jupiter.api.Test"/>
                            </or>
``` holds false too

**Comment by OLibutzki** [comment]
I try to map thesefindings to the different parts of this issue:

> 1. *JavaElementDelegateJunitLaunch* is not consulted for building the editor's context menu
> 1. *JavaElementDelegateJunitLaunch* does not respect `@Testable`
> 1. The context menu for the editor does not respect `@Testable` unless I add the org.junit.jupiter:junit-jupiter-api

`<test property="org.eclipse.jdt.core.hasTypeOnClasspath" value="org.junit.jupiter.api.Test"/>`
Yeah, this explains why I have to add the jupiter-api jar to the classpath. I guess that replacing `org.junit.jupiter.api.Test` with `org.junit.platform.commons.annotation.Testable` solves the third part of this issue.

And this is the block for the editor context menu:
`
         <contextualLaunch>
            <enablement>
               <with variable="selection">
                  <count value="1"/>
               	  <iterate>
               	     <and>
               	       <not><adapt type="org.eclipse.jdt.core.IJavaElement"/></not>
                       <adapt type="org.eclipse.xtext.xbase.ui.launching.JavaElementDelegateJunitLaunch">
                         <adapt type="org.eclipse.jdt.core.IJavaElement">
		                   <test property="org.eclipse.jdt.core.isInJavaProjectWithNature" value="org.eclipse.pde.PluginNature"/>
		                   <test property="org.eclipse.jdt.core.hasTypeOnClasspath" value="junit.framework.Test"/>
		                   <test property="org.eclipse.jdt.junit.canLaunchAsJUnit" forcePluginActivation="true"/>
		                 </adapt>
                       </adapt>
                     </and>
		          </iterate>
               </with>
            </enablement>
         </contextualLaunch>
`
JUnit5 is ignored entirely...

If `@Testable`  is added here as well the issue's first part should be solved.

The second part can be solved by respecting `@Testable` in `JavaElementDelegateJunitLaunch `.

**Comment by cdietrich** [comment]
@OLibutzki for the second part i already created https://github.com/eclipse/xtext-xtend/pull/520

**Comment by OLibutzki** [comment]
@cdietrich, ok, any chance to get the other stuff done in 2.15, too?

**Comment by cdietrich** [comment]
the main problem is that i dont know what the proper solution is: test for api or test for extension etc
and we are busy with other stuff so that i cannot spend 2-3 days to figure out a solution that works properly in all cases

**Comment by cdietrich** [comment]
see https://github.com/eclipse/xtext-eclipse/pull/770 too

**Comment by OLibutzki** [comment]
`@Testable` is the annotation which is intended to use by IDEs. So checking for its existence on the classpath as a first "quick check" is reasonable.

Moreover, `@Testable` needs to be taken into account in `JavaElementDelegateJunitLaunch#containsJUnitTestMethod `. There you have to check for `@Testable` annotated elements and for annotations which are meta-annotated with `@Testable`.

**Comment by cdietrich** [comment]
yes but you can write tests without `@Testable` or with inheriting from that (if i understand that correct) and my junit 5 knlowedge is so sparse that i would need to read on junit 5 for 2 days (the most part of the 2-3 days i estimated)

**Comment by OLibutzki** [comment]
Basically JUnit5 is seperated into two phases: Discovery and execution. In the discovery phase all the tests are collected based on a given DiscoverySelector. A DiscoverySelector might be a class, a method, a package,...

In the execution phase the discovered tests are executed.

As the discovery scan relies on compiled sources and on their execution, IDEs can not use it to determine the executable elements. That's the reason why `@Testable` has been introduced. It provides a chance to identify testable elements at source code level.

Of course this information might differ from the results of the discovery phase, but that's intentional. The JavaDoc of `@Testable` is quite good and worth a read in order to understand the motivation: https://junit.org/junit5/docs/5.2.0/api/org/junit/platform/commons/annotation/Testable.html
